README.md
passenger Cookbook
==================
[![Build Status](https://travis-ci.org/jhx/cookbook-passenger.png?branch=master)](https://travis-ci.org/jhx/cookbook-passenger)
[![Dependency Status](https://gemnasium.com/jhx/cookbook-passenger.png)](https://gemnasium.com/jhx/cookbook-passenger)

Installs and configures passenger and nginx for the Rails app.


Requirements
------------
### Cookbooks
The following cookbooks are direct dependencies because they're used for common "default" functionality.

- `build-essential`
- `logrotate`
- `ohai`
- `rails_app`
- `rvm`
- `yum`

### Platforms
The following platforms are supported and tested under Test Kitchen:

- CentosOS 5.10, 6.5

Other RHEL family distributions are assumed to work. See [TESTING](TESTING.md) for information about running tests in Opscode's Test Kitchen.


Attributes
----------
Refer to [`attributes/default.rb`](attributes/default.rb) for default values.

- `node['passenger']['version']` - version of passenger
- `node['passenger']['version_map']` - options hash to map passenger version to nginx version
- `node['passenger']['nginx']['prefix]` - path to nginx installation directory
- `node['passenger']['nginx']['user]` - nginx user
- `node['passenger']['nginx']['conf_path]` - path to nginx configuration file
- `node['passenger']['nginx']['configure_flags]` - nginx configuration flags applied during compilation
- `node['passenger']['nginx']['modules]` - nginx modules compiled incorporated into binary


Recipes
-------
This cookbook provides one main recipe for configuring a node.

- `default.rb` - *Use this recipe* to install required support packages and gems.
- `nginx.rb` - *Use this recipe* to install and configure passenger and nginx.
- `ohai_plugin.rb` - *Use this recipe* to install an ohai plugin to support the nginx install.

### default
This recipe installs required support packages, plugins, rvm, ruby, and gems.

### nginx
This recipe includes the default recipe, installs the nginx daemon control script, creates log files, configures logrotate, creates nginx user, and installs/configures nginx with the passenger module.

### ohai_plugin
This recipe installs an ohai plugin to support the nginx install.


Usage
-----
On client nodes, use the following recipes:

````javascript
{ "run_list": ["recipe[passenger::nginx]"] }
````

The following are the key items achieved by this cookbook:

- installs required support packages and plugins
- installs rvm, ruby, required gems
- installs nginx daemon control script
- creates log files, installs logrotate configuration file
- creates nginx user, installs/configures nginx with the passenger module


License & Authors
-----------------
- Author:: Doc Walker (<doc.walker@jameshardie.com>)

````text
Copyright 2013-2014, James Hardie Building Products, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
````


CHANGELOG.md
CHANGELOG
=========

0.1.1 / 2014-02-22
------------------

- `NEW` add dependency badge to [README](README.md) | [#51][]
- `NEW` add emeril gem to publish cookbook | [#49][]
- `IMPROVE` update CHANGELOG | [#48][]
- `IMPROVE` - update [README](README.md) | [#47][]
- `IMPROVE` - add missing block closure comments | [#46][]
- `IMPROVE` - update `test-kitchen` gem version | [#44][]
- `IMPROVE` - update spec style (describe v. context, node.set[]=) | [#42][]
- `IMPROVE` - update `rubocop` gem version | [#40][]
- `IMPROVE` - update `chefspec` gem version | [#39][]
- `IMPROVE` - update Gemfile to remove redundant gems | [#38][]
- `NEW` - add [TESTING](TESTING.md) | [#37][]
- `NEW` - add integration tests (via test-kitchen/serverspec) | [#35][]
- `IMPROVE` - require rspec expect syntax | [#32][]
- `IMPROVE` - update .gitignore | [#29][]
- `NEW` - add [CONTRIBUTING](CONTRIBUTING.md) | [#16][]
- `NEW` - add ruby lint checks (via rubocop), fix flagged items | [#15][] [#30][] [#31][]
- `NEW` - add unit tests (via chefspec) | [#14][] [#33][] [#34][]
- `NEW` - add syntax checks (via knife) | [#13][]
- `NEW` - add support for travis-ci | [#12][]
- `NEW` - add cookbook lint checks (via foodcritic), fix flagged items | [#10][] [#11][]
- `NEW` - add specification files for ruby version | [#9][]


0.1.0 / 2013-01-26
------------------

- `NEW` - initial release


Legend
------

- `BREAK`   - breaking changes
- `FIX`     - bug fix
- `IMPROVE` - improvement of existing feature
- `NEW`     - new feature

<!--- The following link definition list is generated by PimpMyChangelog --->
[#9]: https://github.com/jhx/cookbook-passenger/issues/9
[#10]: https://github.com/jhx/cookbook-passenger/issues/10
[#11]: https://github.com/jhx/cookbook-passenger/issues/11
[#12]: https://github.com/jhx/cookbook-passenger/issues/12
[#13]: https://github.com/jhx/cookbook-passenger/issues/13
[#14]: https://github.com/jhx/cookbook-passenger/issues/14
[#15]: https://github.com/jhx/cookbook-passenger/issues/15
[#16]: https://github.com/jhx/cookbook-passenger/issues/16
[#29]: https://github.com/jhx/cookbook-passenger/issues/29
[#30]: https://github.com/jhx/cookbook-passenger/issues/30
[#31]: https://github.com/jhx/cookbook-passenger/issues/31
[#32]: https://github.com/jhx/cookbook-passenger/issues/32
[#33]: https://github.com/jhx/cookbook-passenger/issues/33
[#34]: https://github.com/jhx/cookbook-passenger/issues/34
[#35]: https://github.com/jhx/cookbook-passenger/issues/35
[#37]: https://github.com/jhx/cookbook-passenger/issues/37
[#38]: https://github.com/jhx/cookbook-passenger/issues/38
[#39]: https://github.com/jhx/cookbook-passenger/issues/39
[#40]: https://github.com/jhx/cookbook-passenger/issues/40
[#42]: https://github.com/jhx/cookbook-passenger/issues/42
[#44]: https://github.com/jhx/cookbook-passenger/issues/44
[#46]: https://github.com/jhx/cookbook-passenger/issues/46
[#47]: https://github.com/jhx/cookbook-passenger/issues/47
[#48]: https://github.com/jhx/cookbook-passenger/issues/48
[#49]: https://github.com/jhx/cookbook-passenger/issues/49
[#51]: https://github.com/jhx/cookbook-passenger/issues/51


spec/chef/knife.rb
# encoding: utf-8
cookbook_path '.'
cache_type 'BasicFile'
cache_options(:path => "#{ENV['HOME']}/.chef/checksums")


spec/nginx_spec.rb
# encoding: utf-8
require 'spec_helper'

describe 'passenger::nginx' do
  before do
    # required for travis-ci
    stub_command("bash -c \"source /etc/profile && type rvm | " \
      "cat | head -1 | grep -q '^rvm is a function$'\"").and_return(true)
    stub_command("bash -c \"source /etc/profile.d/rvm.sh && type rvm | " \
      "cat | head -1 | grep -q '^rvm is a function$'\"").and_return(true)
  end # before

  cached(:chef_run) do
    ChefSpec::Runner.new do |node|
      # override cookbook attributes
      node.set['passenger']['nginx']['prefix'] = '/opt/nginx-qa'
      node.set['passenger']['nginx']['user'] = 'nginx-qa'
      node.set['passenger']['ruby_string'] = '1.9.3-fake'
      node.set['passenger']['version'] = '3.0.19-fake'
      node.set['passenger']['version_map']['3.0.19-fake'] = '1.2.6-fake'

      # required because default value is nil
      node.set['rvm']['install_pkgs'] = []

      # required for build-essential cookbook on travis-ci
      node.set['platform_family'] = 'rhel'
    end.converge(described_recipe)
  end # cached

  describe 'passenger' do
    it 'includes described recipe' do
      expect(chef_run).to include_recipe(subject)
    end # it
  end # describe

  describe '/etc/rc.d/init.d/nginx' do
    it 'creates template with expected owner, group, mode' do
      expect(chef_run).to create_template(subject)
        .with(:owner => 'root', :group => 'root', :mode => '0755')
    end # it

    it 'renders file with expected nginx=' do
      expect(chef_run).to render_file(subject)
        .with_content('nginx="/opt/nginx-qa/sbin/nginx"')
    end # it

    it 'renders file with expected NGINX_CONF_FILE=' do
      expect(chef_run).to render_file(subject)
        .with_content('NGINX_CONF_FILE="/opt/nginx-qa/conf/nginx.conf"')
    end # it

    it 'notifies service[nginx]' do
      resource = chef_run.template(subject)
      expect(resource).to notify('service[nginx]').to(:reload).delayed
    end # it
  end # describe

  describe '/var/log/nginx' do
    it 'creates directory with expected owner, group, mode' do
      expect(chef_run).to create_directory(subject)
        .with(:owner => 'root', :group => 'root', :mode => '0755')
    end # it
  end # describe

  describe '/etc/logrotate.d/nginx' do
    it 'creates template with expected owner, group, mode' do
      expect(chef_run).to create_template(subject)
        .with(:owner => 'root', :group => 'root', :mode => '0644')
    end # it

    it 'renders file with expected path' do
      expect(chef_run).to render_file(subject)
        .with_content('/var/log/nginx/*.log')
    end # it

    it 'renders file with expected frequency' do
      expect(chef_run).to render_file(subject)
        .with_content('daily')
    end # it

    it 'renders file with expected rotation limit' do
      expect(chef_run).to render_file(subject)
        .with_content('rotate 30')
    end # it

    it 'renders file with expected options (missingok)' do
      expect(chef_run).to render_file(subject)
        .with_content('missingok')
    end # it

    it 'renders file with expected options (compress)' do
      expect(chef_run).to render_file(subject)
        .with_content('compress')
    end # it

    it 'renders file with expected options (delaycompress)' do
      expect(chef_run).to render_file(subject)
        .with_content('delaycompress')
    end # it

    it 'renders file with expected options (sharedscripts)' do
      expect(chef_run).to render_file(subject)
        .with_content('sharedscripts')
    end # it

    it 'renders file with expected postrotate command' do
      expect(chef_run).to render_file(subject).with_content(
        '[ ! -f /var/run/nginx.pid ] || kill -USR1 `cat /var/run/nginx.pid`'
      )
    end # it
  end # describe

  it 'creates user nginx-qa' do
    expect(chef_run).to create_user('nginx-qa')
      .with(:system => true, :shell => '/bin/nologin', :home => '/var/www')
  end # it

  # TODO: write spec for rvm_shell

  it 'enables service nginx' do
    expect(chef_run).to enable_service('nginx')
  end # it

  it 'starts service nginx' do
    expect(chef_run).to start_service('nginx')
  end # it

  describe '/opt/nginx-qa/conf/nginx.conf' do
    it 'creates template with expected owner, group, mode' do
      expect(chef_run).to create_template(subject)
        .with(:owner => 'root', :group => 'root', :mode => '0644')
    end # it

    it 'renders file with expected passenger_root' do
      expect(chef_run).to render_file(subject)
        .with_content('passenger_root ' \
        '/usr/local/rvm/gems/1.9.3-fake/gems/passenger-3.0.19-fake;')
    end # it

    it 'renders file with expected passenger_ruby' do
      expect(chef_run).to render_file(subject)
        .with_content('passenger_ruby ' \
        '/usr/local/rvm/wrappers/1.9.3-fake/ruby;')
    end # it

    it 'renders file with expected passenger_base_uri neo' do
      expect(chef_run).to render_file(subject)
        .with_content(/passenger_base_uri\s.*\/neo;/)
    end # it

    it 'renders file with expected passenger_base_uri morpheus' do
      expect(chef_run).to render_file(subject)
        .with_content(/passenger_base_uri\s.*\/morpheus;/)
    end # it

    it 'renders file with expected passenger_pre_start neo' do
      expect(chef_run).to render_file(subject)
        .with_content(%r{passenger_pre_start\s.*http://localhost/neo;})
    end # it

    it 'renders file with expected passenger_pre_start morpheus' do
      expect(chef_run).to render_file(subject)
        .with_content(%r{passenger_pre_start\s.*http://localhost/morpheus;})
    end # it

    it 'renders file with expected rewrite' do
      expect(chef_run).to render_file(subject)
        .with_content('rewrite ' \
        '^/$ $scheme://$http_host/neo permanent;')
    end # it

    it 'renders file with expected location /neo' do
      expect(chef_run).to render_file(subject)
        .with_content('location /neo {')
    end # it

    it 'renders file with expected location /morpheus' do
      expect(chef_run).to render_file(subject)
        .with_content('location /morpheus {')
    end # it

    it 'notifies service[nginx]' do
      resource = chef_run.template(subject)
      expect(resource).to notify('service[nginx]').to(:reload).delayed
    end # it
  end # describe

end # describe


spec/spec_helper.rb
# encoding: utf-8
require 'chefspec'
require 'chefspec/berkshelf'
require 'chefspec/cacher'
Dir.glob(File.dirname(__FILE__) + '/helpers/**/*', &method(:require))

RSpec.configure do |config|
  config.expect_with :rspec do |c|
    c.syntax = :expect
  end # config.expect_with
end # RSpec


recipes/default.rb
# encoding: utf-8
#
# Cookbook Name:: passenger
# Recipe:: default
#
# Copyright 2013, James Hardie Building Products, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

#--------------------------------------------------- include dependent recipes
include_recipe 'build-essential'

include_recipe 'passenger::ohai_plugin'

include_recipe 'rails_app'

include_recipe 'rvm'
include_recipe 'rvm::system'
include_recipe 'rvm::gem_package'

#------------------------------------------------ install package dependencies
%w(ruby-devel curl-devel).each do |pkg|
  package pkg
end # %w(...)

#------------------------------------------------------ install passenger ruby
rvm_environment node['passenger']['ruby_string']

#---------------------------------------------------- install gem dependencies
rvm_global_gem 'rake'
rvm_global_gem 'ohai'
rvm_global_gem 'passenger' do
  version node['passenger']['version']
  action :install
end # rvm_global_gem


recipes/ohai_plugin.rb
# encoding: utf-8
#
# Cookbook Name:: passenger
# Recipe:: ohai_plugin
#
# Author:: Jamie Winsor (<jamie@vialstudios.com>)
# Author:: Doc Walker (<doc.walker@jameshardie.com>)
#
# Copyright 2012, Riot Games
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

ohai 'reload_passenger_nginx' do
  action :nothing
  plugin 'passenger'
end # ohai

template "#{node['ohai']['plugin_path']}/passenger.rb" do
  source    'plugins/passenger.rb.erb'
  owner     'root'
  group     'root'
  mode      '0755'
  variables(
    :passenger_nginx_prefix => node['passenger']['nginx']['prefix']
  )
  notifies  :reload, 'ohai[reload_passenger_nginx]', :immediately
end # template

include_recipe 'ohai'


recipes/nginx.rb
# encoding: utf-8
#
# Cookbook Name:: passenger
# Recipe:: nginx
#
# Copyright 2013, James Hardie Building Products, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'passenger'

#--------------------------------------------------- install nginx init script
template '/etc/rc.d/init.d/nginx' do # or is it /etc/rc.d/init.d/nginx?
  owner     'root'
  group     'root'
  mode      '0755'
  notifies  :reload, 'service[nginx]'
end # template

#------------------------------------------------- prepare nginx log directory
directory '/var/log/nginx' do
  owner     'root'
  group     'root'
  mode      '0755'
  recursive true
end # directory

#------------------------------------------------------- configure logrotate.d
logrotate_app 'nginx' do
  cookbook      'logrotate'
  template_mode '0644'
  path          '/var/log/nginx/*.log'
  frequency     'daily'
  rotate        30
  options       %w(missingok compress delaycompress sharedscripts)
  postrotate    '[ ! -f /var/run/nginx.pid ] || ' \
    'kill -USR1 `cat /var/run/nginx.pid`'
end # logrotate_app

#----------------------------------------------------------- create nginx user
user node['passenger']['nginx']['user'] do
  system  true
  shell   '/bin/nologin'
  home    '/var/www'
end # user

#---------------------------------------------- install passenger nginx module
passenger_root = "/usr/local/rvm/gems/#{node['passenger']['ruby_string']}" +
  "/gems/passenger-#{node['passenger']['version']}"
passenger_ruby = '/usr/local/rvm/wrappers/' \
  "#{node['passenger']['ruby_string']}/ruby"

nginx_signature = {
  'version' => node['passenger']['version_map']
    .fetch(node['passenger']['version']),
  'passenger_version' => node['passenger']['version'],
  'prefix' => node['passenger']['nginx']['prefix'],
  'ruby_string' => node['passenger']['ruby_string'],
  'with' => [
    'http_ssl_module',
    'cc-opt=-Wno-error',
    *node['passenger']['nginx']['modules']
    ].sort
}

extra_configure_flags = node['passenger']['nginx']['modules']
  .map { |flag| "--with-#{flag}" }

configure_flags = node['passenger']['nginx']['configure_flags']
  .map { |flag| "--#{flag}" }

rvm_shell "#{passenger_root}/bin/passenger-install-nginx-module" do
  ruby_string node['passenger']['ruby_string']
  code [
    "#{passenger_root}/bin/passenger-install-nginx-module",
    *configure_flags,
    "--extra-configure-flags=\"#{extra_configure_flags.join(' ')}\""
  ].join(' ')
  not_if do
    node.automatic_attrs.fetch('passenger') { {} }
      .fetch('nginx_signature') { '' } == nginx_signature
  end
  notifies :restart, 'service[nginx]'
  # /opt/nginx/sbin/nginx -t # test configuration and exit
end # rvm_shell

service 'nginx' do
  supports :status => true, :restart => true, :reload => true
  action [:enable, :start]
end # service

#-------------------------------------------- install nginx configuration file
# TODO: need to fail if passenger_root or passenger_ruby is nil
# TODO: try rvm_shell verify version for above?
template node['passenger']['nginx']['conf_path'] do
  owner     'root'
  group     'root'
  mode      '0644'
  variables :passenger_root => passenger_root,
            :passenger_ruby => passenger_ruby
  notifies  :reload, 'service[nginx]'
end # template
